File: src/main/java/com/morak/performancetracker/context/ResultComposite.java
package com.morak.performancetracker.context;

import java.util.List;

public class ResultComposite implements Result {

    private TestMetadata testMetadata;
    private List<Result> subResults;

    public ResultComposite(TestMetadata testMetadata, List<Result> subResults) {
        this.testMetadata = testMetadata;
        this.subResults = subResults;
    }

    @Override
    public void findAndAdd(Result result, TestMetadata testMetadata) {
        if (this.testMetadata.getClassName().equals(testMetadata.getClassName())) {
            this.subResults.add(result);
        }
        subResults.forEach(subResult -> subResult.findAndAdd(result, testMetadata));
    }

    @Override
    public String getResult() {
        return this.testMetadata.toString();
    }

    @Override
    public List<Result> getSubResults() {
        return subResults;
    }

    public void add(ResultComposite resultComposite) {
        this.subResults.add(resultComposite);
    }
}

File: src/main/kotlin/com/morak/performancetracker/context/ResultComposite.kt
package com.morak.performancetracker.context


class ResultComposite(
    private val testMetadata: TestMetadata,
    subResults: List<Result> = emptyList()
) : Result {
    private val _subResults: MutableList<Result> = subResults.toMutableList()
    override val subResults: List<Result>
        get() = _subResults.toList()

    override val result: String
        get() = testMetadata.self

    override fun findAndAdd(result: Result, testMetadata: TestMetadata) {
        if (this.testMetadata == testMetadata) {
            _subResults.add(result)
            return
        }

        subResults.forEach { it.findAndAdd(result, testMetadata) }
    }
}
